src/goTypeDefinition.ts

```diff
@@ -0,0 +1,96 @@
+/*---------------------------------------------------------
+ * Copyright (C) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------*/
+
+'use strict';
+
+import vscode = require('vscode');
+import cp = require('child_process');
+import path = require('path');
+import { byteOffsetAt, getBinPath, canonicalizeGOPATHPrefix, getFileArchive, killTree } from './util';
+import { promptForMissingTool } from './goInstallTools';
+import { getToolsEnvVars } from './util';
+
+interface GuruDescribeOutput {
+	desc: string;
+	pos: string;
+	detail: string;
+	value: GuruDescribeValueOutput;
+}
+
+interface GuruDescribeValueOutput {
+	type: string;
+	value: string;
+	objpos: string;
+	typespos: GuruDefinitionOutput[];
+}
+
+interface GuruDefinitionOutput {
+	objpos: string;
+	desc: string;
+}
+
+export class GoTypeDefinitionProvider implements vscode.TypeDefinitionProvider {
+	provideTypeDefinition(document: vscode.TextDocument, position: vscode.Position, token: vscode.CancellationToken): vscode.ProviderResult<vscode.Definition> {
+		return new Promise<vscode.Location[]>((resolve, reject) => {
+			let goGuru = getBinPath('guru');
+			if (!path.isAbsolute(goGuru)) {
+				promptForMissingTool('guru');
+				return reject('Cannot find tool "guru" to find type definitions.');
+			}
+
+			let filename = canonicalizeGOPATHPrefix(document.fileName);
+			let cwd = path.dirname(filename);
+			let offset = byteOffsetAt(document, position);
+			let env = getToolsEnvVars();
+			let buildTags = vscode.workspace.getConfiguration('go', document.uri)['buildTags'];
+			let args = buildTags ? ['-tags', buildTags] : [];
+			args.push('-json', '-modified', 'describe', `${filename}:#${offset.toString()}`);
+
+			let process = cp.execFile(goGuru, args, { env }, (err, stdout, stderr) => {
+				try {
+					if (err && (<any>err).code === 'ENOENT') {
+						promptForMissingTool('guru');
+						return resolve(null);
+					}
+
+					if (err) {
+						return reject(err);
+					}
+
+					let guruOutput = <GuruDescribeOutput>JSON.parse(stdout.toString());
+					let results: vscode.Location[] = [];
+
+					if (!guruOutput.value || !guruOutput.value.typespos) {
+						if (guruOutput.value) {
+							console.log('No type definitions found: probably you have not latest version of guru');
+						}
+						return resolve(null);
+					}
+
+					guruOutput.value.typespos.forEach(ref => {
+						let match = /^(.*):(\d+):(\d+)/.exec(ref.objpos);
+						if (!match) return;
+						let [_, file, lineStartStr, colStartStr] = match;
+						let referenceResource = vscode.Uri.file(path.resolve(cwd, file));
```

Comment:
You are considering the directory of the file to be the `cwd` which is set above to be the directory where the file we invoke this on is. But you run guru with the working directory of the extension host. If guru does output relative paths in some cases this can cause a failure to calculate the correct path. guru seems to output absolute paths for me though.

Reply:
Removed resolve, since I think guru will always return absolute paths.